File: src/main/java/destiny/data/gauquelin/BirthAndDeathReader.java
/**
 * @author smallufo 
 * Created on 2008/3/10 at 下午 4:35:39
 */ 
package destiny.data.gauquelin;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.net.URL;

import destiny.core.calendar.Location;
import destiny.core.calendar.Time;
import destiny.core.calendar.Location.EastWest;
import destiny.core.calendar.Location.NorthSouth;
import destiny.core.calendar.eightwords.EightWords;
import destiny.core.calendar.eightwords.EightWordsContextBean;
import destiny.utils.location.TimeZoneUtils;

public class BirthAndDeathReader
{
  public BirthAndDeathReader()
  {
    EightWordsContextBean bean;
    
    URL url = getClass().getResource("BirthAndDeath.txt");
    File file;
    BufferedReader bReader = null;
    
    try
    {
      file = new File(url.toURI());
      FileReader fReader = new FileReader(file);
      bReader = new BufferedReader(fReader);
      String line = null;
      Time birthTime = null;
      Time deathTime = null;
      
      int i=0;
      while ((line = bReader.readLine()) != null)
      {
        if (!line.equals("") && line.charAt(0) != '#')
        {
          char birthOrDeath = line.charAt(5);
          int day = Integer.valueOf(line.substring(11, 13).trim());
          int month = Integer.valueOf(line.substring(13, 15).trim());
          int year = Integer.valueOf(line.substring(15, 19).trim());
          int hour = Integer.valueOf(line.substring(19, 21).trim());
          int minute = Integer.valueOf(line.substring(21, 23).trim());
          int second = Integer.valueOf(line.substring(23, 25).trim());
          int tz = Integer.valueOf(line.substring(25, 27).trim());
          
          int latDeg = Integer.valueOf(line.substring(27, 29).trim());
          char NS = line.charAt(29);
          int latMin = Integer.valueOf(line.substring(30, 32).trim());
          
          int longDeg = Integer.valueOf(line.substring(32, 34).trim());
          char EW = line.charAt(34);
          int longMin = Integer.valueOf(line.substring(35, 37).trim());
          
          
          Time time = new Time(year , month , day , hour , minute , second);
          Location location = new Location( (EW == 'E' ? EastWest.EAST : EastWest.WEST) , longDeg , longMin , 0 , (NS == 'N' ? NorthSouth.NORTH : NorthSouth.SOUTH) , latDeg , latMin , 0 , 0 , TimeZoneUtils.getTimeZone(tz*60) );
          
          if (birthOrDeath == 'B')
            birthTime = time;
          else
          {
            deathTime = time;
            //計算存活多久
            if (deathTime.diffSeconds(birthTime) < 60*60*24*365)
            {
              i++;
              bean = new EightWordsContextBean();
              EightWords ew1 = bean.getEightWords(birthTime , location);
              System.out.println("======"+i+"=====");
              System.out.print("出生於 : " + ew1);

              bean = new EightWordsContextBean();
              EightWords ew2 = bean.getEightWords(deathTime , location);
              System.out.println("\n卒於 : " + ew2);
            }
          }
        }
      }//while
    }
    catch (Exception e)
    {
      e.printStackTrace();
    }
  }
  
  public static void main(String[] args)
  {
    new BirthAndDeathReader();
  }
}

File: src/main/java/destiny/data/gauquelin/BirthAndDeathReader.kt
/**
 * Created by smallufo on 2018-04-06.
 */
package destiny.data.gauquelin

import destiny.core.calendar.Location
import java.io.InputStream
import java.time.Duration
import java.time.LocalDateTime

/**
 * "9280" (i.e. 9272) Birth and Death Data (mostly infants) in Paris (unpublished)
 *
 * 9272 Birth and Death Data (mostly infants) in Paris. "Series F, volume 1".
 * 讀取的是 http://cura.free.fr/gauq/902gdF1.html
 * 此為 2009 改版的格式 , 用 TAB 區隔欄位
 * 沒有性別！
 * 有四筆資料 Death 「早於」 Birth , txt 檔案已經 mark 起來
 * */

class BirthAndDeathReader {

  fun read() {
    val iStream: InputStream = javaClass.getResourceAsStream("BirthAndDeath.txt")
    iStream.bufferedReader(Charsets.UTF_8).useLines { sequence ->
      sequence
        .filterNot { it.startsWith("#") }
        .filterNot { it.length == 0 }
        .chunked(2).forEach { lines ->
          val birth = parseLine(lines[0])
          val death = parseLine(lines[1])
          require(death.lmt.isAfter(birth.lmt))
          println("[${birth.number}] 存活 ${Duration.between(birth.lmt , death.lmt).toDays()} 日")

        }
    }
  }

  fun parseLine(line : String) : GPerson2 {
    val tokens = line.split("\t").toList()
    val num = tokens[0].toInt()
    val day = tokens[2].toInt()
    val month = tokens[3].toInt()
    val year = tokens[4].toInt()
    val hour = tokens[5].toInt()
    val min = tokens[6].toInt()
    val sec = tokens[7].toInt()
    val ciCode = tokens[8]

    val (tzid , minuteOffset) = ParseTools.getTzidAndMinuteOffset(tokens[9].toInt())
    val lat = ParseTools.parseLat(tokens[10])
    val lng = ParseTools.parseLng(tokens[11])

    val lmt = LocalDateTime.of(year , month , day , hour , min , sec)
    val loc = Location(lat, lng, tzid, minuteOffset)

    return GPerson2(null , num , null , null , lmt , loc , null , line)
  }

}
